project/npda/general_functions/visit_categories.py

```diff
@@ -1,28 +1,85 @@
     """
     Returns visit categories present in this visit instance, and tags them as to whether they contain errors
     """
     categories = []
-    for category in VISIT_FIELDS:
-        category_present = False
-        category_error_present = False
-        for field in category[1]:
-            if hasattr(visit_instance, field):
-                if getattr(visit_instance, field) is not None:
-                    category_present = True
-                    errors = visit_instance.errors
-                    if errors:
-                        if field in errors:
-                            category_error_present = True
-        categories.append(
-            {
-                "category": category[0].value,
-                "present": category_present,
-                "has_error": category_error_present,
-            }
-        )
+
+    for _, tab in VISIT_CATEGORIES_BY_TAB.items():
+        for category_name, category in tab.items():
+            fields = category["fields"]
+
+            present = False
+            errors = {}
+
+            # Data can be either:
+            #  - On the bound form field after submitting the questionnaire
+            if form:
+                for field in form:
+                    if field.name in fields:
+                        present = True
+                        
+                        if field.errors:
+                            errors[field.name] = field.errors
+
+            #  - On the instance itself after a CSV upload
+            if instance:
+                for field in fields:
+                    if getattr(instance, field):
+                        present = True
+
+                if instance.errors:
+                    for field in instance.errors.keys():
+                        if field in fields:
+                            errors[field] = [error["message"] for error in instance.errors[field]]
+
+            categories.append(
+                {
+                    "name": category_name,
+                    "present": present,
+                    "errors": errors,
+                    "anchor": urllib.parse.quote_plus(category_name),
+                    "colour": category["colour"]
+                }
+            )
+
     return categories
 
+
+def get_visit_tabs(form):
+    tabs = []
+    instance = form.instance if form else None
+
+    all_categories = get_visit_categories(instance, form)
+
+    assigned_active_tab = False
+
+    for tab_name, categories in VISIT_CATEGORIES_BY_TAB.items():
+        category_names = categories.keys()
+        categories = [c for c in all_categories if c["name"] in category_names]
+
+        errors = {}
+        for category in categories:
+            for field, field_errors in category["errors"].items():
+                errors[field] = field_errors
+
+        tab = {
+            "name": tab_name,
+            "categories": categories,
+            "errors": errors
+        }
+
+        # Show the first tab with errors
```

Comment:
The javascript helper also does this and scrolls to the right place but can't hurt to have the HTML render as expected anyway